Documentation improvements (#466)
Amsterdam/dso-api: dev-docs/source/streaming.rst

* The rendering classes perform ``json.dumps()`` calls on single records.
* The ``Response`` class is replaced by a ``StreamingResponse`` class.

Chunked JSON Rendering
----------------------

The JSON response is written by selectively applying ``json.dumps()``.
The idea is to write some object layout manually, and leverage ``json.dumps()`` where possible.
It basically looks like:

.. code-block:: python

    yield '{["_embedded": ['

    # write main listing:
    for record in generator:
        if not first:
            yield ",\n"
        yield json.dumps(record)

    yield "],\n"

    # write "_links" and "page" parts without the opening/closing braces:
    yield json.dumps(footer_links)[1:-1]
    yield "}\n"

Additional Optimizations
------------------------

Output Buffering
~~~~~~~~~~~~~~~~

To avoid too many back/forth calls from the response-generator
and the WSGI server, the produced output is submitted in chunks of 4096kB.
Otherwrite an OS ``write()`` call might happen for a simple ``yield "}"`` statement.

Next Link Optimization
~~~~~~~~~~~~~~~~~~~~~~

A standard paginator would do an expensive ``COUNT(*)`` on the table

> and use that to tell whether thare are additional pages.

Reviewer: and use that to tell whether _there_ are additional pages.

Reviewer 2: This is digging into implementation details far too deeply. Just say that determining the number of pages is too expensive. That's true for the remotes as well.

We've optimized this by requesting one extra record from the database at the end of a page.

> This sentinel record is not rendered. It's existence does does indicate there is another page available.

Reviewer: ```suggestion
This sentinel record is not rendered. It's existence indicates there is another page available.
```

Reviewer 2: ```suggestion
This sentinel record is not rendered. Its existence indicates that there is another page available.
```

This does mean the next/previous links have to be written at end of the response,
after all main objects have been seen. This is handled by writing the main JSON object in chunks.

Fields like a "total page count" and "total results" are no longer available, but rarely needed either.
Typically clients only need a link to the next page.

> 

Reviewer: ```suggestion
And we support the "_count" query parameter for when a client does need a result count. 
```

Author: tx, added!
